servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/ConnectionObserverInitializer.java

```diff
@@ -155,22 +149,68 @@ static final class ConnectionObserverHandler extends ChannelDuplexHandler {
         public void handlerAdded(final ChannelHandlerContext ctx) {
             final Channel channel = ctx.channel();
             if (channel.isActive()) {
+                maybeAddChannelClosedListener(channel);
                 reportTcpHandshakeComplete(channel);
                 if (handshakeOnActive) {
                     reportSecurityHandshakeStarting(sslConfig);
                 }
             }
         }
 
+        @Override
+        public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress, SocketAddress localAddress,
+                            ChannelPromise promise) throws Exception {
+            // The connect promise is the first to be notified of problems in the connect phase, even before the
+            // channel.closeFuture(). We typically want to know about closure before we get to returning the failed
+            // response because it is the cause of the failed response, so we eagerly propagate that in the event of
+            // a failed connect promise. We add our listener to the promise before forwarding the call so we can be
+            // sure this callback is fired early in the failure pathway.
+            promise.addListener((ChannelFuture future) -> {
+                if (future.isSuccess()) {
+                    maybeAddChannelClosedListener(ctx.channel());
+                } else {
+                    addedCloseListener = true;
+                    Throwable t = channelError(future.channel());
+                    if (t == null) {
+                        t = future.cause();
+                    }
+                    if (t == null) {
+                        observer.connectionClosed();
+                    } else {
+                        observer.connectionClosed(t);
+                    }
+                }
+            });
+            super.connect(ctx, remoteAddress, localAddress, promise);
+        }
+
         @Override
         public void channelActive(final ChannelHandlerContext ctx) {
-            reportTcpHandshakeComplete(ctx.channel());
+            Channel channel = ctx.channel();
+            maybeAddChannelClosedListener(channel);
+            reportTcpHandshakeComplete(channel);
             if (handshakeOnActive) {
                 reportSecurityHandshakeStarting(sslConfig);
             }
```

Comment:
Consider moving these lines to a new method (something like `onChannelActive(Channel)` to avoid duplication between this method and `handlerAdded`